src/rust/src/wallet.rs

```diff
@@ -135,33 +155,113 @@ impl Wallet {
+    pub fn is_checkpointed(&self) -> bool {
+        self.last_checkpoint.is_some()
+    }
+
+    /// Rewinds the note commitment tree to the last checkpoint. If the rewind operation
+    /// succeeds, this also decrements the checkpoint height.
+    pub fn rewind(&mut self, to_height: BlockHeight) -> bool {
+        if let Some(checkpoint_height) = self.last_checkpoint.as_mut() {
+            // if the rewind length exceeds the number of checkpoints we have stored,
+            // return failure.
+            let mut blocks_to_rewind: i64 =
+                <i64>::from(to_height) - <i64>::from(*checkpoint_height);
+            if blocks_to_rewind > self.witness_tree.checkpoints().len().try_into().unwrap() {
+                return false;
+            }
+
+            // when we rewind, we need to remove the spentness marks for the notes we're tracking
+            while blocks_to_rewind > 0 {
+                if !self.witness_tree.rewind() {
+                    return false;
+                }
+
+                *checkpoint_height = *checkpoint_height - 1;
+                for (txid, n) in self.wallet_tx_notes.iter() {
+                    if n.tx_height > *checkpoint_height {
+                        self.spent_notes.retain(|o| o.txid != *txid)
+                    }
+                }
+                // remove any notes that were mined since the last checkpoint
+                self.wallet_tx_notes
+                    .retain(|_, n| n.tx_height <= *checkpoint_height);
+                // reset our last observed height to ensure that notes added in the future are
+                // from a new block
+                self.last_observed = Some(LastObserved {
+                    block_height: *checkpoint_height,
+                    block_tx_idx: None,
+                });
+
+                blocks_to_rewind -= 1;
+            }
+
+            true
+        } else {
+            false
+        }
     }
 
     /// Add note data for those notes that are decryptable with one of this wallet's
-    /// incoming viewing keys to the wallet, and return the indices of the actions
-    /// that we were able to decrypt.
+    /// incoming viewing keys to the wallet, and return the indices of the actions that we
+    /// were able to decrypt.
     pub fn add_notes_from_bundle(
         &mut self,
+        block_height: BlockHeight,
         txid: &TxId,
         bundle: &Bundle<Authorized, Amount>,
     ) -> Vec<usize> {
         let mut tx_notes = TxNotes {
+            tx_height: block_height,
             decrypted_notes: BTreeMap::new(),
         };
 
+        // Check for spends of our notes by matching against the nullifiers
+        // we're tracking, and when we detect one, associate the current
+        // txid as spending the note.
+        for (action_idx, action) in bundle.actions().iter().enumerate() {
+            let ord_nf = action.nullifier();
+            if let Some(txids) = self.nullifiers.get_mut(ord_nf) {
```

Comment:
Yes, in fact, this will be common. We're iterating all the actions of the bundle, and checking every action's nullifier against the set of nullifiers that correspond to notes we've received. So it will be `None` in every case *except* in the case that the action is a spend of one of our notes.